**复习：**这门课程得主要目的是通过真实的数据，以实战的方式了解数据分析的流程和熟悉数据分析python的基本操作。知道了课程的目的之后，我们接下来我们要正式的开始数据分析的实战教学，完成kaggle上[泰坦尼克的任务](https://www.kaggle.com/c/titanic/overview)，实战数据分析全流程。
这里有两份资料：
教材《Python for Data Analysis》和 baidu.com &
google.com（善用搜索引擎）

## 1 第一章：数据载入及初步观察

### 1.1 载入数据
数据集下载 https://www.kaggle.com/c/titanic/overview

#### 1.1.1 任务一：导入numpy和pandas

In [1]:
import numpy as np
import pandas as pd

#### 1.1.2 任务二：载入数据
(1) 使用相对路径载入数据  
(2) 使用绝对路径载入数据

In [4]:
df = pd.read_csv('train.csv')
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [6]:
#写入代码
df = pd.read_csv('C://Users//shepherd//Desktop//hands-on-data-analysis/train.csv')
df.head()

#使用os.getcwd()前需要导入os库
import os
os.getcwd()

'C:\\Users\\shepherd\\Desktop\\hands-on-data-analysis'

【提示】相对路径载入报错时，尝试使用os.getcwd()查看当前工作目录。 

【思考】知道数据加载的方法后，试试pd.read_csv()和pd.read_table()的不同，如果想让他们效果一样，需要怎么做？了解一下'.tsv'和'.csv'的不同，如何加载这两个数据集？  
【总结】加载的数据是所有工作的第一步，我们的工作会接触到不同的数据格式（eg:.csv;.tsv;.xlsx）,但是加载的方法和思路都是一样的，在以后工作和做项目的过程中，遇到之前没有碰到的问题，要多多查资料吗，使用google，了解业务逻辑，明白输入和输出是什么。

In [7]:
#对比pd.read_csv()和pd.read_table()的不同
df = pd.read_csv('train.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [8]:
#用pd.read_table导入文件并输出,发现df2输出结果不整齐
df2 = pd.read_table('train.csv')
print(type(df2))
df2.head()

<class 'pandas.core.frame.DataFrame'>


PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0  1,0,3,"Braund, Mr. Owen Harris",male,22,1,0,A/...                             
1  2,1,1,"Cumings, Mrs. John Bradley (Florence Br...                             
2  3,1,3,"Heikkinen, Miss. Laina",female,26,0,0,S...                             
3  4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May ...                             
4  5,0,3,"Allen, Mr. William Henry",male,35,0,0,3...

In [9]:
#从https://www.geek-share.com/detail/2772467139.html中知道两个函数间的区别，主要区别在于分别分隔符上
df2 = pd.read_table('train.csv', sep = ',')
df2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

#### 1.1.3 任务三：每1000行为一个数据模块，逐块读取

In [10]:
#写入代码
chunker = pd.read_csv('train.csv', chunksize = 1000)
print(type(chunker))
for piece in chunker:
    print(type(piece))
    print(len(piece))
    print(piece.head())

<class 'pandas.io.parsers.TextFileReader'>
<class 'pandas.core.frame.DataFrame'>
891
   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113

【思考】什么是逐块读取？为什么要逐块读取呢？
<div class="alert alert-block alert-success">
【回答】对于数据量较大的文件，可考虑进行逐块读取处理，可以减少内存消耗、提高效率等。在pandas中可通过chunksize设置块的大小，可通过对其遍历进行操作；对于大文件，可以使用nrows参数读取部分 </div>

#### 1.1.4 任务四：将表头改成中文，索引改为乘客ID [对于某些英文资料，我们可以通过翻译来更直观的熟悉我们的数据]
PassengerId => 乘客ID  
Survived    => 是否幸存   
Pclass      => 乘客等级(1/2/3等舱位)  
Name        => 乘客姓名  
Sex         => 性别                 
Age         => 年龄                 
SibSp       => 堂兄弟/妹个数  
Parch       => 父母与小孩个数  
Ticket      => 船票信息             
Fare        => 票价                
Cabin       => 客舱                
Embarked    => 登船港口             

In [11]:
#写入代码
df.columns = ['乘客ID','是否幸存','仓位等级','姓名','性别','年龄','兄弟姐妹个数','父母子女个数','船票信息','票价','客舱','登船港口']
df = df.set_index(df['乘客ID'])
df.head()

乘客ID  是否幸存  仓位等级                                                 姓名  \
乘客ID                                                                        
1        1     0     3                            Braund, Mr. Owen Harris   
2        2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
3        3     1     3                             Heikkinen, Miss. Laina   
4        4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
5        5     0     3                           Allen, Mr. William Henry   

          性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
乘客ID                                                                      
1       male  22.0       1       0         A/5 21171   7.2500   NaN    S  
2     female  38.0       1       0          PC 17599  71.2833   C85    C  
3     female  26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
4     female  35.0       1       0            113803  53.1000  C123    S  
5       male  35.0       0       0            373450   8.0500   NaN    S

【思考】所谓将表头改为中文其中一个思路是：将英文额度表头替换成中文。还有其他的方法吗？
<div class="alert alert-block alert-success">
【回答】可以在读取文件时直接修改表头，具体代码为：df = pd.read_csv('train.csv', names=['乘客ID','是否幸存','仓位等级','姓名','性别','年龄','兄弟姐妹个数','父母子女个数','船票信息','票价','客舱','登船港口'],index_col='乘客ID',header=0)；同时有对部分列重命名的方法，其语法为df.rename(columns={‘a’: ‘A’})，其中'a'为原列名，'A'为新列名 </div>

### 1.2 初步观察
导入数据后，你可能要对数据的整体结构和样例进行概览，比如说，数据大小、有多少列，各列都是什么格式的，是否包含null等

#### 1.2.1 任务一：查看数据的基本信息

In [12]:
#写入代码
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   乘客ID    891 non-null    int64  
 1   是否幸存    891 non-null    int64  
 2   仓位等级    891 non-null    int64  
 3   姓名      891 non-null    object 
 4   性别      891 non-null    object 
 5   年龄      714 non-null    float64
 6   兄弟姐妹个数  891 non-null    int64  
 7   父母子女个数  891 non-null    int64  
 8   船票信息    891 non-null    object 
 9   票价      891 non-null    float64
 10  客舱      204 non-null    object 
 11  登船港口    889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 90.5+ KB


乘客ID        是否幸存        仓位等级          年龄      兄弟姐妹个数      父母子女个数  \
count  891.000000  891.000000  891.000000  714.000000  891.000000  891.000000   
mean   446.000000    0.383838    2.308642   29.699118    0.523008    0.381594   
std    257.353842    0.486592    0.836071   14.526497    1.102743    0.806057   
min      1.000000    0.000000    1.000000    0.420000    0.000000    0.000000   
25%    223.500000    0.000000    2.000000   20.125000    0.000000    0.000000   
50%    446.000000    0.000000    3.000000   28.000000    0.000000    0.000000   
75%    668.500000    1.000000    3.000000   38.000000    1.000000    0.000000   
max    891.000000    1.000000    3.000000   80.000000    8.000000    6.000000   

               票价  
count  891.000000  
mean    32.204208  
std     49.693429  
min      0.000000  
25%      7.910400  
50%     14.454200  
75%     31.000000  
max    512.329200

【提示】有多个函数可以这样做，你可以做一下总结

<div class="alert alert-block alert-warning"> 
【回答】目前只知道info和describe两个方法，有待整理 </div>

#### 1.2.2 任务二：观察表格前10行的数据和后15行的数据

In [13]:
df.head(10)

乘客ID  是否幸存  仓位等级                                                 姓名  \
乘客ID                                                                        
1        1     0     3                            Braund, Mr. Owen Harris   
2        2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
3        3     1     3                             Heikkinen, Miss. Laina   
4        4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
5        5     0     3                           Allen, Mr. William Henry   
6        6     0     3                                   Moran, Mr. James   
7        7     0     1                            McCarthy, Mr. Timothy J   
8        8     0     3                     Palsson, Master. Gosta Leonard   
9        9     1     3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
10      10     1     2                Nasser, Mrs. Nicholas (Adele Achem)   

          性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
乘客ID                                                                      
1       male  22.0       1       0         A/5 21171   7.2500   NaN    S  
2     female  38.0       1       0          PC 17599  71.2833   C85    C  
3     female  26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
4     female  35.0       1       0            113803  53.1000  C123    S  
5       male  35.0       0       0            373450   8.0500   NaN    S  
6       male   NaN       0       0            330877   8.4583   NaN    Q  
7       male  54.0       0       0             17463  51.8625   E46    S  
8       male   2.0       3       1            349909  21.0750   NaN    S  
9     female  27.0       0       2            347742  11.1333   NaN    S  
10    female  14.0       1       0            237736  30.0708   NaN    C

In [14]:
df.tail(15)

乘客ID  是否幸存  仓位等级                                             姓名      性别  \
乘客ID                                                                            
877    877     0     3                  Gustafsson, Mr. Alfred Ossian    male   
878    878     0     3                           Petroff, Mr. Nedelio    male   
879    879     0     3                             Laleff, Mr. Kristo    male   
880    880     1     1  Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female   
881    881     1     2   Shelley, Mrs. William (Imanita Parrish Hall)  female   
882    882     0     3                             Markun, Mr. Johann    male   
883    883     0     3                   Dahlberg, Miss. Gerda Ulrika  female   
884    884     0     2                  Banfield, Mr. Frederick James    male   
885    885     0     3                         Sutehall, Mr. Henry Jr    male   
886    886     0     3           Rice, Mrs. William (Margaret Norton)  female   
887    887     0     2                          Montvila, Rev. Juozas    male   
888    888     1     1                   Graham, Miss. Margaret Edith  female   
889    889     0     3       Johnston, Miss. Catherine Helen "Carrie"  female   
890    890     1     1                          Behr, Mr. Karl Howell    male   
891    891     0     3                            Dooley, Mr. Patrick    male   

        年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
乘客ID                                                              
877   20.0       0       0              7534   9.8458   NaN    S  
878   19.0       0       0            349212   7.8958   NaN    S  
879    NaN       0       0            349217   7.8958   NaN    S  
880   56.0       0       1             11767  83.1583   C50    C  
881   25.0       0       1            230433  26.0000   NaN    S  
882   33.0       0       0            349257   7.8958   NaN    S  
883   22.0       0       0              7552  10.5167   NaN    S  
884   28.0       0       0  C.A./SOTON 34068  10.5000   NaN    S  
885   25.0       0       0   SOTON/OQ 392076   7.0500   NaN    S  
886   39.0       0       5            382652  29.1250   NaN    Q  
887   27.0       0       0            211536  13.0000   NaN    S  
888   19.0       0       0            112053  30.0000   B42    S  
889    NaN       1       2        W./C. 6607  23.4500   NaN    S  
890   26.0       0       0            111369  30.0000  C148    C  
891   32.0       0       0            370376   7.7500   NaN    Q

#### 1.2.4 任务三：判断数据是否为空，为空的地方返回True，其余地方返回False

In [15]:
#写入代码
df.isnull().head()

乘客ID   是否幸存   仓位等级     姓名     性别     年龄  兄弟姐妹个数  父母子女个数   船票信息     票价  \
乘客ID                                                                           
1     False  False  False  False  False  False   False   False  False  False   
2     False  False  False  False  False  False   False   False  False  False   
3     False  False  False  False  False  False   False   False  False  False   
4     False  False  False  False  False  False   False   False  False  False   
5     False  False  False  False  False  False   False   False  False  False   

         客舱   登船港口  
乘客ID                
1      True  False  
2     False  False  
3      True  False  
4     False  False  
5      True  False

【总结】上面的操作都是数据分析中对于数据本身的观察

【思考】对于一个数据，还可以从哪些方面来观察？找找答案，这个将对下面的数据分析有很大的帮助

<div class="alert alert-block alert-success">
【回答】可以关注数据类型，对于数值型数据可以计算其数字特性和绘制直方图等掌握数据的分布等基本情况 </div>

### 1.3 保存数据

#### 1.3.1 任务一：将你加载并做出改变的数据，在工作目录下保存为一个新文件train_chinese.csv

In [16]:
df.to_csv('train_chinese.csv')

**复习：**数据分析的第一步，加载数据我们已经学习完毕了。当数据展现在我们面前的时候，我们所要做的第一步就是认识他，今天我们要学习的就是**了解字段含义以及初步观察数据**。

### 1.4 知道你的数据叫什么
我们学习pandas的基础操作，那么上一节通过pandas加载之后的数据，其数据类型是什么呢？

#### 1.4.1 任务一：pandas中有两个数据类型DateFrame和Series，通过查找简单了解他们。然后自己写一个关于这两个数据类型的小例子🌰[开放题]

In [9]:
# 序列数据
sdata = {'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}
example_1 = pd.Series(sdata)
example_1

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

In [7]:
# 数据框
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003],'pop': [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
example_2 = pd.DataFrame(data)
example_2

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

#### 1.4.2 任务二：根据上节课的方法载入"train.csv"文件


In [47]:
df = pd.read_csv('train.csv')

也可以加载上一节课保存的"train_chinese.csv"文件。通过翻译版train_chinese.csv熟悉了这个数据集，然后我们对trian.csv来进行操作
#### 1.4.3 任务三：查看DataFrame数据的每列的项

In [24]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

#### 1.4.4任务四：查看"cabin"这列的所有项 [有多种方法]

In [30]:
# 方法一
df['Cabin'].head(5)

0     NaN
1     C85
2     NaN
3    C123
4     NaN
Name: Cabin, dtype: object

In [31]:
# 方法二
df.Cabin.head(5)

0     NaN
1     C85
2     NaN
3    C123
4     NaN
Name: Cabin, dtype: object

#### 1.4.5 任务五：加载文件"test_1.csv"，然后对比"train.csv"，看看有哪些多出的列，然后将多出的列删除
经过我们的观察发现一个测试集test_1.csv有一列是多余的，我们需要将这个多余的列删去

In [35]:
test_1 = pd.read_csv('test_1.csv')
test_1.head()

Unnamed: 0  PassengerId  Survived  Pclass  \
0           0            1         0       3   
1           1            2         1       1   
2           2            3         1       3   
3           3            4         1       1   
4           4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked    a  
0      0         A/5 21171   7.2500   NaN        S  100  
1      0          PC 17599  71.2833   C85        C  100  
2      0  STON/O2. 3101282   7.9250   NaN        S  100  
3      0            113803  53.1000  C123        S  100  
4      0            373450   8.0500   NaN        S  100

In [36]:
del test_1['a']
test_1.head(5)

Unnamed: 0  PassengerId  Survived  Pclass  \
0           0            1         0       3   
1           1            2         1       1   
2           2            3         1       3   
3           3            4         1       1   
4           4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

【思考】还有其他的删除多余的列的方式吗？

<div class="alert alert-block alert-success">
【回答】法二代码：test_1.drop(['a'],axis=1) </div>


#### 1.4.6 任务六： 将['PassengerId','Name','Age','Ticket']这几个列元素隐藏，只观察其他几个列元素

In [39]:
df.drop(['PassengerId','Name','Age','Ticket'],axis=1).head()

Survived  Pclass     Sex  SibSp  Parch     Fare Cabin Embarked
0         0       3    male      1      0   7.2500   NaN        S
1         1       1  female      1      0  71.2833   C85        C
2         1       3  female      0      0   7.9250   NaN        S
3         1       1  female      1      0  53.1000  C123        S
4         0       3    male      0      0   8.0500   NaN        S

【思考】对比任务五和任务六，是不是使用了不一样的方法(函数)，如果使用一样的函数如何完成上面的不同的要求呢？

<div class="alert alert-block alert-success">
 【回答】从结果上来看，显示部分列和删除无关列的是相同的，比如任务六可以在导入时就选择所需列，代码为：pd.read_csv('test_1.csv',usecols=['Survived', 'Pclass', 'Sex', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked']).head(5)
</div>

【思考回答】

如果想要完全的删除你的数据结构，使用inplace=True，因为使用inplace就将原数据覆盖了，所以这里没有用

<div class="alert alert-block alert-success">
【回答】首先，drop只是隐藏某几列，原数据结构保持不变，仅是展示部分内容。而inplace=True将原数据覆盖，删除了原来的数据结构，保留的是删除某几列之后的数据;inplace=False时原数据不变。
</div>


### 1.5 筛选的逻辑

表格数据中，最重要的一个功能就是要具有可筛选的能力，选出我所需要的信息，丢弃无用的信息。

下面我们还是用实战来学习pandas这个功能。

#### 1.5.1 任务一： 我们以"Age"为筛选条件，显示年龄在10岁以下的乘客信息。

In [40]:
df[df['Age']<10].head()

PassengerId  Survived  Pclass                                      Name  \
7             8         0       3            Palsson, Master. Gosta Leonard   
10           11         1       3           Sandstrom, Miss. Marguerite Rut   
16           17         0       3                      Rice, Master. Eugene   
24           25         0       3             Palsson, Miss. Torborg Danira   
43           44         1       2  Laroche, Miss. Simonne Marie Anne Andree   

       Sex  Age  SibSp  Parch         Ticket     Fare Cabin Embarked  
7     male  2.0      3      1         349909  21.0750   NaN        S  
10  female  4.0      1      1        PP 9549  16.7000    G6        S  
16    male  2.0      4      1         382652  29.1250   NaN        Q  
24  female  8.0      3      1         349909  21.0750   NaN        S  
43  female  3.0      1      2  SC/Paris 2123  41.5792   NaN        C

#### 1.5.2 任务二： 以"Age"为条件，将年龄在10岁以上和50岁以下的乘客信息显示出来，并将这个数据命名为midage

In [49]:
midage = df[(df['Age']>10)&(df['Age']<50)]
midage.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

【提示】了解pandas的条件筛选方式以及如何使用交集和并集操作

#### 1.5.3 任务三：将midage的数据中第100行的"Pclass"和"Sex"的数据显示出来

In [43]:
midage.loc[[100],['Pclass','Sex']]

Pclass     Sex
100       3  female

In [51]:
midage = midage.reset_index(drop = True)
midage.loc[[100],['Pclass','Sex']]

Pclass   Sex
100       2  male

【思考】这个reset_index()函数的作用是什么？如果不用这个函数，下面的任务会出现什么情况？

<div class="alert alert-block alert-success">
【回答】不使用reset_index()函数的结果为第一结果，对比发现，是索引的区别，该函数的具体解释如下:rop=True 就是把原来的索引index列去掉，重置index；drop=False 就是保留原来的索引，添加重置的index。两者的区别就是有没有把原来的index去掉。
</div>


#### 1.5.4 任务四：将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [52]:
midage.loc[[100,105,108],['Pclass','Sex','Name']]

Pclass   Sex                               Name
100       2  male  Byles, Rev. Thomas Roussel Davids
105       3  male           Cribb, Mr. John Hatfield
108       3  male                    Calic, Mr. Jovo

【提示】使用pandas提出的简单方式，你可以看看loc方法

对比整体的数据位置，你有发现什么问题吗？那么如何解决？

#### 1.5.5 任务五：使用iloc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [53]:
midage.iloc[[100,105,108],[2,3,4]] #无法用列名索引吗

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

<div class="alert alert-block alert-success">
【总结】首先区分下loc()和iloc()方法的区别，前者是根据索引名称确定数据，后者是根据数据位置确定数据。此时loc和iloc输出的结果相同，原因是数据框所用的索引“乘客ID”是从1开始计数的。若索引值是从0开始计数，那么iloc()方法的索引名称为100时，实际定位到101行数据。
</div>


**复习：**在前面我们已经学习了Pandas基础，知道利用Pandas读取csv数据的增删查改，今天我们要学习的就是**探索性数据分析**，主要介绍如何利用Pandas进行排序、算术计算以及计算描述函数describe()的使用。

In [54]:
#载入之前保存的train_chinese.csv数据，关于泰坦尼克号的任务，我们就使用这个数据
text = pd.read_csv('train_chinese.csv')
text.head()

乘客ID  乘客ID.1  是否幸存  仓位等级  \
0     1       1     0     3   
1     2       2     1     1   
2     3       3     1     3   
3     4       4     1     1   
4     5       5     0     3   

                                                  姓名      性别    年龄  兄弟姐妹个数  \
0                            Braund, Mr. Owen Harris    male  22.0       1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0       1   
2                             Heikkinen, Miss. Laina  female  26.0       0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0       1   
4                           Allen, Mr. William Henry    male  35.0       0   

   父母子女个数              船票信息       票价    客舱 登船港口  
0       0         A/5 21171   7.2500   NaN    S  
1       0          PC 17599  71.2833   C85    C  
2       0  STON/O2. 3101282   7.9250   NaN    S  
3       0            113803  53.1000  C123    S  
4       0            373450   8.0500   NaN    S

### 1.6 了解你的数据吗？
教材《Python for Data Analysis》第五章

#### 1.6.1 任务一：利用Pandas对示例数据进行排序，要求升序

In [2]:
# 具体请看《利用Python进行数据分析》第五章 排序和排名 部分
#自己构建一个都为数字的DataFrame数据
frame = pd.DataFrame(np.arange(8).reshape((2, 4)),index = ['three', 'one'], columns = ['d', 'a', 'b', 'c'])
frame

d  a  b  c
three  0  1  2  3
one    4  5  6  7

【代码解析】

pd.DataFrame() ：创建一个DataFrame对象 

np.arange(8).reshape((2, 4)) : 生成一个二维数组（2*4）,第一列：0，1，2，3 第二列：4，5，6，7

index=['2, 1] ：DataFrame 对象的索引列

columns=['d', 'a', 'b', 'c'] ：DataFrame 对象的索引行

【问题】：大多数时候我们都是想根据列的值来排序,所以将你构建的DataFrame中的数据根据某一列，升序排列

In [4]:
frame.sort_values(by='a')

d  a  b  c
three  0  1  2  3
one    4  5  6  7

【思考】通过书本你能说出Pandas对DataFrame数据的其他排序方式吗？

<div class="alert alert-block alert-success">
【回答】pandas中排序的方法有sort_index()和sort_value()。
</div>


【总结】下面将不同的排序方式做一个总结

1.让行索引升序排序

In [5]:
frame.sort_index()

d  a  b  c
one    4  5  6  7
three  0  1  2  3

2.让列索引升序排序


In [6]:
frame.sort_index(axis=1)

a  b  c  d
three  1  2  3  0
one    5  6  7  4

3.让列索引降序排序

In [7]:
frame.sort_index(axis=1, ascending=False)

d  c  b  a
three  0  3  2  1
one    4  7  6  5

4.让任选两列数据同时降序排序

In [10]:
frame.sort_values(by=['a', 'b'], ascending=False) 

d  a  b  c
one    4  5  6  7
three  0  1  2  3

#### 1.6.2 任务二：对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列），从这个数据中你可以分析出什么？

In [14]:
'''
在开始我们已经导入了train_chinese.csv数据，而且前面我们也学习了导入数据过程，根据上面学习，我们直接对目标列进行排序即可
head(20) : 读取前20条数据

'''
text = pd.read_csv('train_chinese.csv')
text.sort_values(by=['票价', '年龄'], ascending=False).head(20)

乘客ID  乘客ID.1  是否幸存  仓位等级  \
679   680     680     1     1   
258   259     259     1     1   
737   738     738     1     1   
438   439     439     0     1   
341   342     342     1     1   
88     89      89     1     1   
27     28      28     0     1   
742   743     743     1     1   
311   312     312     1     1   
299   300     300     1     1   
118   119     119     0     1   
380   381     381     1     1   
716   717     717     1     1   
700   701     701     1     1   
557   558     558     0     1   
527   528     528     0     1   
377   378     378     0     1   
779   780     780     1     1   
730   731     731     1     1   
689   690     690     1     1   

                                                    姓名      性别    年龄  兄弟姐妹个数  \
679                 Cardeza, Mr. Thomas Drake Martinez    male  36.0       0   
258                                   Ward, Miss. Anna  female  35.0       0   
737                             Lesurer, Mr. Gustave J    male  35.0       0   
438                                  Fortune, Mr. Mark    male  64.0       1   
341                     Fortune, Miss. Alice Elizabeth  female  24.0       3   
88                          Fortune, Miss. Mabel Helen  female  23.0       3   
27                      Fortune, Mr. Charles Alexander    male  19.0       3   
742              Ryerson, Miss. Susan Parker "Suzette"  female  21.0       2   
311                         Ryerson, Miss. Emily Borie  female  18.0       2   
299    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0       0   
118                           Baxter, Mr. Quigg Edmond    male  24.0       0   
380                              Bidois, Miss. Rosalie  female  42.0       0   
716                      Endres, Miss. Caroline Louise  female  38.0       0   
700  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female  18.0       1   
557                                Robbins, Mr. Victor    male   NaN       0   
527                                 Farthing, Mr. John    male   NaN       0   
377                          Widener, Mr. Harry Elkins    male  27.0       0   
779  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...  female  43.0       0   
730                      Allen, Miss. Elisabeth Walton  female  29.0       0   
689                  Madill, Miss. Georgette Alexandra  female  15.0       0   

     父母子女个数      船票信息        票价               客舱 登船港口  
679       1  PC 17755  512.3292      B51 B53 B55    C  
258       0  PC 17755  512.3292              NaN    C  
737       0  PC 17755  512.3292             B101    C  
438       4     19950  263.0000      C23 C25 C27    S  
341       2     19950  263.0000      C23 C25 C27    S  
88        2     19950  263.0000      C23 C25 C27    S  
27        2     19950  263.0000      C23 C25 C27    S  
742       2  PC 17608  262.3750  B57 B59 B63 B66    C  
311       2  PC 17608  262.3750  B57 B59 B63 B66    C  
299       1  PC 17558  247.5208          B58 B60    C  
118       1  PC 17558  247.5208          B58 B60    C  
380       0  PC 17757  227.5250              NaN    C  
716       0  PC 17757  227.5250              C45    C  
700       0  PC 17757  227.5250          C62 C64    C  
557       0  PC 17757  227.5250              NaN    C  
527       0  PC 17483  221.7792              C95    S  
377       2    113503  211.5000              C82    C  
779       1     24160  211.3375               B3    S  
730       0     24160  211.3375               B5    S  
689       1     24160  211.3375               B5    S

【思考】排序后，如果我们仅仅关注年龄和票价两列。根据常识我知道发现票价越高的应该客舱越好，所以我们会明显看出，票价前20的乘客中存活的有14人，这是相当高的一个比例，那么我们后面是不是可以进一步分析一下票价和存活之间的关系，年龄和存活之间的关系呢？当你开始发现数据之间的关系了，数据分析就开始了。

当然，这只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

**多做几个数据的排序**

In [23]:
text.sort_values(by=['年龄', '仓位等级']).head(20)

乘客ID  乘客ID.1  是否幸存  仓位等级                                 姓名      性别  \
803   804     804     1     3    Thomas, Master. Assad Alexander    male   
755   756     756     1     2          Hamalainen, Master. Viljo    male   
469   470     470     1     3      Baclini, Miss. Helene Barbara  female   
644   645     645     1     3             Baclini, Miss. Eugenie  female   
78     79      79     1     2      Caldwell, Master. Alden Gates    male   
831   832     832     1     2    Richards, Master. George Sibley    male   
305   306     306     1     1     Allison, Master. Hudson Trevor    male   
183   184     184     1     2          Becker, Master. Richard F    male   
827   828     828     1     2              Mallet, Master. Andre    male   
164   165     165     0     3       Panula, Master. Eino Viljami    male   
172   173     173     1     3       Johnson, Miss. Eleanor Ileen  female   
381   382     382     1     3        Nakid, Miss. Maria ("Mary")  female   
386   387     387     0     3    Goodwin, Master. Sidney Leonard    male   
788   789     789     1     3         Dean, Master. Bertram Vere    male   
297   298     298     0     1       Allison, Miss. Helen Loraine  female   
340   341     341     1     2     Navratil, Master. Edmond Roger    male   
530   531     531     1     2           Quick, Miss. Phyllis May  female   
7       8       8     0     3     Palsson, Master. Gosta Leonard    male   
16     17      17     0     3               Rice, Master. Eugene    male   
119   120     120     0     3  Andersson, Miss. Ellis Anna Maria  female   

       年龄  兄弟姐妹个数  父母子女个数             船票信息        票价       客舱 登船港口  
803  0.42       0       1             2625    8.5167      NaN    C  
755  0.67       1       1           250649   14.5000      NaN    S  
469  0.75       2       1             2666   19.2583      NaN    C  
644  0.75       2       1             2666   19.2583      NaN    C  
78   0.83       0       2           248738   29.0000      NaN    S  
831  0.83       1       1            29106   18.7500      NaN    S  
305  0.92       1       2           113781  151.5500  C22 C26    S  
183  1.00       2       1           230136   39.0000       F4    S  
827  1.00       0       2  S.C./PARIS 2079   37.0042      NaN    C  
164  1.00       4       1          3101295   39.6875      NaN    S  
172  1.00       1       1           347742   11.1333      NaN    S  
381  1.00       0       2             2653   15.7417      NaN    C  
386  1.00       5       2          CA 2144   46.9000      NaN    S  
788  1.00       1       2        C.A. 2315   20.5750      NaN    S  
297  2.00       1       2           113781  151.5500  C22 C26    S  
340  2.00       1       1           230080   26.0000       F2    S  
530  2.00       1       1            26360   26.0000      NaN    S  
7    2.00       3       1           349909   21.0750      NaN    S  
16   2.00       4       1           382652   29.1250      NaN    Q  
119  2.00       4       2           347082   31.2750      NaN    S

<div class="alert alert-block alert-success">
仿照上文思路，对年龄和客舱等级进行排序，也有14人存活，这一定程度上展现了对婴幼儿的关怀，具体规律有待进一步探究
</div>

#### 1.6.3 任务三：利用Pandas进行算术计算，计算两个DataFrame数据相加结果

In [16]:
# 具体请看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分

#自己构建两个都为数字的DataFrame数据
frame_a = pd.DataFrame(np.arange(9.).reshape((3, 3)), columns=list('bcd'),index=['Ohio', 'Texas', 'Colorado'])
frame_b = pd.DataFrame(np.arange(12.).reshape((4, 3)), columns=list('bde'),index=['Utah', 'Ohio', 'Texas', 'Oregon'])

将frame_a和frame_b进行相加


In [17]:
frame_a + frame_b

b   c     d   e
Colorado  NaN NaN   NaN NaN
Ohio      3.0 NaN   6.0 NaN
Oregon    NaN NaN   NaN NaN
Texas     9.0 NaN  12.0 NaN
Utah      NaN NaN   NaN NaN

【提醒】两个DataFrame相加后，会返回一个新的DataFrame，对应的行和列的值会相加，没有对应的会变成空值NaN。<br>
当然，DataFrame还有很多算术运算，如减法，除法等，有兴趣的同学可以看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分，多在网络上查找相关学习资料。

#### 1.6.4 任务四：通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？

In [24]:
'''
还是用之前导入的chinese_train.csv如果我们想看看在船上，最大的家族有多少人（‘兄弟姐妹个数’+‘父母子女个数’），我们该怎么做呢？
'''
max(text['兄弟姐妹个数'] + text['父母子女个数'])
max(text['兄弟姐妹个数']) + max(text['父母子女个数'])

14

【提醒】我们只需找出”兄弟姐妹个数“和”父母子女个数“之和最大的数，当然你还可以想出很多方法和思考角度，欢迎你来说出你的看法。
<div class="alert alert-block alert-warning"> 
【回答】暂时没想出比较好的方法 </div>



#### 1.6.5 任务五：学会使用Pandas describe()函数查看数据基本统计信息

In [19]:
#(1) 关键知识点示例做一遍（简单数据）
# 具体请看《利用Python进行数据分析》第五章 汇总和计算描述统计 部分

#自己构建一个有数字有空值的DataFrame数据
frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

调用 describe 函数，观察frame2的数据基本信息

In [21]:
frame2.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

#### 1.6.6 任务六：分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据，你能发现什么？

In [22]:
'''
看看泰坦尼克号数据集中 票价 这列数据的基本统计数据
'''
text['票价'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: 票价, dtype: float64

【思考】从上面数据我们可以看出，试试在下面写出你的看法。然后看看我们给出的答案。

当然，答案只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

【回答】可以推测数据为右偏分布

**多做几个组数据的统计，看看你能分析出什么？**

In [25]:
text['年龄'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: 年龄, dtype: float64

<div class="alert alert-block alert-success">
【思考】一共有714个票价数据，其平均值为29.699118，标准差为14.526497，证明票价波动比较大，25%的人票价低于20，50%低于28，75%低于38，票价最大值为80，最小值为0.42。
</div>

【总结】本节中我们通过Pandas的一些内置函数对数据进行了初步统计查看，这个过程最重要的不是大家得掌握这些函数，而是看懂从这些函数出来的数据，构建自己的数据分析思维，这也是第一章最重要的点，希望大家学完第一章能对数据有个基本认识，了解自己在做什么，为什么这么做，后面的章节我们将开始对数据进行清洗，进一步分析。

【总结】数据的加载以及入门，接下来就要接触数据本身的运算，我们将主要掌握numpy和pandas在工作和项目场景的运用。